### ISSUE TITLE
Submission: GROUP 15: chocolate_exploration

### ISSUE THREAD

**Issue by manvirsingh96** [UBC-MDS/data-analysis-review-2022#20]
**Submitting authors:**  @YXIN15 @netsgnut @manvirsingh96

**Repo Link: https://github.com/UBC-MDS/chocolate_exploration:**
**Report Link :  https://github.com/UBC-MDS/chocolate_exploration/blob/main/doc/chocolate_exploration_results_report.pdf**

**Abstract/executive summary:** 

Dark chocolate is an aquired taste. Not every dark chocolate will be received well. As three avid chocolate lovers, we decided to see how a particular dark chocolate will be rated using different factors like the country of manufacure, the ingredients in the chocolate, the proportion of cacao etc.
Our project involves using a regression approach to predict the rating of a dark chocolate based on certain factors. This can help new manufacturers predict if a new chocolate manufactured will be rated well in the market. We use regression models like KNN, Ridge, SVM RBF and Decision Tree in an attempt to make these predictions.

**Editor:** @flor14 
**Reviewer:** Jonah Hamilton, Ranjit Sundaramurthi, Waiel Tinwala 

- [x] I agree to abide by [MDS's Code of Conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/) during the review process and in maintaining my package should it be accepted.

**Comment by xXJohamXx** [comment]
## Data analysis review checklist

### Reviewer: <GITHUB_USERNAME> 

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [x] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [ ] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robsutness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [x] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [ ] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [x] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [x] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 1.2

### Review Comments: 

Please provide more detailed feedback here on what was done particularly well, and what could be improved. It is especially important to elaborate on items that you were not able to check off in the list above.

1) Great work team! I really enjoyed reading through your project. Your repository is organized well and it is easy for the reader to move through the steps of the analysis that you outlined in the .README. One small thing would be to add your affiliations in the final report as this is listed as a requirement under the analysis report section above. also, there are some small grammatical errors in the report: after the comma in the second paragraph of the introduction How should not be capitalized. present in the next sentence should be changed to presented. Aide should be aid. There are other instances where the grammar could be cleaned up. It obviously does not impact the technical aspect of the project but in terms of communication/delivery of results I would recommend using Grammarly or something similar to check the document. 

2) The categorical plots do a good job at communicating the overall distribution but the labels on the y-axis are very close together and difficult to read. Maybe it would be possible to increase the axis/image size, so the labels are easier to read. Also, for the numeric plots I think adding some color could be helpful as the rest of the report is in black and white.

3) something that I’m interested in is what attributes are associated with a high rating in the dataset? Perhaps you could include some sort of visualization that shows what features are associated with a high rating, or to include a visualization on the distribution of ratings in the dataset.

4) The results of the hyperparameter tuning are clearly displayed and are explained well but there is no mention of the max test features row. I'm assuming this relates to the cross-validation results of using Count Vectorizer but I think this warrants mentioning in the analysis/results section.

5) Table 3 clearly summarizes the MAPE scores of all the models tested and clearly shows that the SVM model preforms the best. Another idea that could give more insight into your project and that is linked to your introduction where you mention that the model could be used to determine some important factors in the development of chocolate, would be to include something regarding the most influence data points in the final model.

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by ranjitprakash1986** [comment]
## Data analysis review checklist

### Reviewer: @ranjitprakash1986

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [ ] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [ ] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [x] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [ ] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robsutness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [ ] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [ ] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [x] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [ ] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 2

### Review Comments: 

I would like to commend the project team for selecting an interesting yet challenging task of predicting chocolate ratings through Machine Learning.  The project was engaging to read and understand how the data was cleansed and prepared for the modelling. The modelling process is an open field and there are several ways of approaching the goal. It was interesting to read that journey of the team members through the EDA and Project reports. 

Here are some improvements that can be made to further solidify the project in the upcoming Milestones:

1. The dependencies are well documents and I successfully created the environment as prescribed. The installation for R packages did not execute successfully. This can be made robust to ensure that the commands in the Usage section run universally without any issues.
	
2. The Readme file could be treated as the last document to be updated to reflect the latest status of the project. There were mentions of the Make file which is in progress and could lead the reader to assume that those are completed and ready to use. A good practice for next milestone might be to ensure the Readme is consistent with the repository status.
	
3. The  models have been run on the training and validation data.  The MAPE of 8.9% is attractive. A mention about what is the target MAPE in the context of this Use-Case will put the current metrics in context of the final goal. I could not find the model performance on test data. I am curious to see if the performance ranking for the algorithms would change when they are applied on the unseen test data.

4. I liked the choice of the ML algorithms and the delineation of the hyper-parameters relevant to each algorithm in Table 1. An explanation on inferences drawn on the models with respect to their performance will be nice to include in the final report. For example what makes the Decision Tree the worst performing model as compared to the best performing model SVM RBF.  Enlisting he limitations of each model such as the Decision Tree being more prone to overfitting etc. will give the reader a more comprehensive understanding of the analysis.
	
5. The selection of the scoring metric is intriguing. The motivation behind the choice of the scoring metric as MAPE is something the readers will be interested to understand. There might be a typographical error on the report which mentions "negative MAPE" as the metric. However the tables and rest of the analysis are using MAPE.
	
6. It is interesting to note that "review_date" is used as a feature and treated as a numeric datatype. There is a temporal aspect of using date as a feature while modeling.  It will be intriguing to see how this feature affects the predictions when treated as a numeric datatype. Since ratings can change over time, the shuffling feature of "train_test_splits" can be questionable. It inherently assumes that the data is random and no trend is existent in the ordering of the data. Any trends based on the "review_date" might be lost with a shuffled split. This might be complicated to solve and we haven't encountered such a scenario in any MDS classes yet. However, I wanted to make a comment to capture this aspect of the project and hopefully we will learn to tackle such use cases in the upcoming blocks.

The project looks in good shape and I look forward to viewing the completed project in the upcoming Milestones. Nice work!

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by WaielonH** [comment]
## Data analysis review checklist

### Reviewer: Waiel Tinwala @WaielonH 

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [x] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robsutness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [ ] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [ ] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [x] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [x] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 2

### Review Comments: 

First of all, Amazing work team! congratulations on making it this far. The project is very well made and all around checks most of the important boxes. A few comments that I'd like to make are written below. Note that these comments have been made in much less time than I would have liked to, so do understand if there's anything that I might have misunderstood. 

1. The report is well written and is easy to understand. Other than a few grammatical mistakes, the only gripe I have with the report is that all the links are non-clickable. I believe this is a minute change which significantly improves readability and also provides easier access to the provided resources. Barring these small issues, the report effectively communicates the entire project in a single document.
2. While there is an extensive and in depth guide to reproduce the analysis on any machine, the automation needs some improvement. I was unable to run the scripts without independently installing certain packages and libraries (pandoc, magick). As a result of some packages not installing, other scripts down the line break.
3. Taking a closer look at the EDA charts, the plots manage to convey the desired interpretation but I found that I had to zoom in quite a bit sometimes to understand what is going on. The data in the plot can be enhanced much more with a few extra design choices such as highlighting certain peaks, making the text more readable, perhaps including density plots for some of the columns. A little bit of color goes a long way.
4. The project organization is done exceptionally here. the scripts are commented and have been modularized, every document is organized neatly into relevant folders making it significantly easier to navigate through.
5. A bit more on the choice of plots. While I do agree with the choice of charts that is already provided, I believe something like a correlation plot between the target variable and the relevant features would be very interesting to see. This would show any linear relationship between the features and target and could also perhaps help in feature selection down the line.

I wish the team all the best for the future and would be delighted to see where the project goes from here.

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by YXIN15** [comment]
Improvements Based on Feedback
**EDA**
- [x] **Adding more elaboration for the EDA plots -- interpretations, insight on the data, conclusions: https://github.com/UBC-MDS/chocolate_exploration/commit/37a2897f1356018f2cda3a2088a69bed90319ff0
- [x]  **Increasing plot-size and adding color to the EDA plots to make it easier to read: https://github.com/UBC-MDS/chocolate_exploration/commit/37a2897f1356018f2cda3a2088a69bed90319ff0
- [x] **Including a visualization that shows what features are associated with a high rating-- perhaps a correlation plot between the target variable and the relevant features to show linear relationships/help with feature selection later on: https://github.com/UBC-MDS/chocolate_exploration/commit/37a2897f1356018f2cda3a2088a69bed90319ff0

**Models**
- [x] Include output in train_test_split. R usage: https://github.com/UBC-MDS/chocolate_exploration/commit/37a2897f1356018f2cda3a2088a69bed90319ff0
- [x] Adding random forest to the set of models, as tree-based ensemble methods typically lead to better results on these types of datasets : https://github.com/UBC-MDS/chocolate_exploration/commit/74e2326855aa95748bc9c3bed620a6da6d480631 
- [x] Fixing issues with the pandoc and magick packages using docker: https://github.com/UBC-MDS/chocolate_exploration/commit/84c272ffe85b26a71a2056210b315d491c9ff94d
- [x] Adding tests to check the inputs (type, size, etc.) and provide error messages if an input is missing, etc: https://github.com/UBC-MDS/chocolate_exploration/commit/c348a8db09820ba3898bc33579f4eb47f79a8b37

**Report**
- [x] Cleaning up the language usage in the final report, adding more background/summary: https://github.com/UBC-MDS/chocolate_exploration/commit/ae430eea18ee0c900824c001bb493fd1cbdb6acd
- [x] Including a target MAPE for the context of this Use-Case to put the current metrics in context of the final goal, motivation for using MAPE as the target: https://github.com/UBC-MDS/chocolate_exploration/commit/b1b6b01c4a2b49e4bc0daef7d1c7bf828c5e9bfe
- [x] Including a max features row in the table of hyperparameters for the Count Vectorizer https://github.com/UBC-MDS/chocolate_exploration/commit/f672e7642f75553c2be09e3b4cf2a8282e0bef12
- [x] Commenting on the model performances and their limitations: https://github.com/UBC-MDS/chocolate_exploration/commit/6260454dbbae6e57a7cf6f7d1f84059db5ece646
